File: qe/data2015.hs

data BST = BEmpty | BLeaf Int | BNode Int BST BST deriving (Show)

t1 = BNode 20 (BNode 11 (BLeaf 5) (BLeaf 17)) (BNode 38 (BLeaf 34) (BLeaf 42))

postOrder BEmpty            = []
postOrder (BLeaf t)         = [t]
postOrder (BNode t bl br)   = postOrder bl ++ postOrder br ++ [t]

addBST i BEmpty = BLeaf i
addBST i (BLeaf n)
    | i <= n = BNode n (BLeaf i) BEmpty
    | otherwise =  BNode n BEmpty (BLeaf i)
addBST i (BNode n bl br)
    | i <= n    = BNode n (addBST i bl) br
    | otherwise = BNode n bl (addBST i br)

topV (BLeaf i) = i
topV (BNode i _ _) = i

checkBST BEmpty = True 
checkBST (BLeaf n) = True 
checkBST (BNode n BEmpty BEmpty) = True 
checkBST (BNode n bl BEmpty)
    | topV bl <= n = True 
    | otherwise = False 
checkBST (BNode n BEmpty br)
    | topV br >= n = True 
    | otherwise = False 
checkBST (BNode n bl br)
    | topV bl <= n && topV br >= n = True 
    | otherwise  = False 

rebuild [] = BEmpty
rebuild [x] = BLeaf x
rebuild xs = BNode lastOne (rebuild leftV) (rebuild rightV)
    where
        lastOne = last xs
        leftV = [x | x <- xs , x < lastOne]
        rightV = [x | x <- xs , x > lastOne]

File: qe/data2015.ml
(* a ocaml version 
   Yihao Sun
   Syracuse
   2021 *)

type int_bst = 
  | BEmpty
  | BLeaf of int
  | BNode of int * int_bst * int_bst;;


let t1 = BNode (20, BNode (11, BLeaf 5, BLeaf 17), BNode (38, BLeaf 34, BLeaf 42));;


let rec post_order t =
  match t with
  | BEmpty -> []
  | BLeaf v -> [v]
  | BNode (v, left, right) -> (post_order left) @ (post_order right) @ [v]

let rec add_bst t insert_v =
  match t with
  | BEmpty -> BLeaf insert_v
  | BLeaf v -> 
    if insert_v > v 
    then BNode (v, (BLeaf insert_v), BEmpty) 
    else BNode (v, BEmpty, (BLeaf insert_v))
  | BNode (v, left, right) ->
    if insert_v > v
    then BNode (v, (add_bst left insert_v), right)
    else BNode (v, left, (add_bst right insert_v));;

let last lst = List.nth lst ((List.length lst) - 1);;

let rec rebuild lst =
    match lst with
    | [] ->  BEmpty
    | [x] -> BLeaf x
    | x :: xs ->
    let last_one = (last lst) in
    let left = rebuild (List.filter ((>) last_one) lst) in
    let right = rebuild (List.filter ((<) last_one) lst) in
    BNode (last_one, left, right);;

rebuild (post_order t1)
